File: python/examples/extraction/01_css_extraction_sdk.py
#!/usr/bin/env python3
"""
CSS Extraction with SDK - No LLM Cost

This example shows how to extract structured data using CSS selectors.
CSS extraction is fast, reliable, and has no LLM cost.

Usage:
    python 01_css_extraction_sdk.py

Requirements:
    pip install crawl4ai-cloud
"""

import asyncio
import json
from crawl4ai_cloud import AsyncWebCrawler, CrawlerRunConfig

# Configuration
API_KEY = "YOUR_API_KEY"  # Replace with your API key


async def extract_with_css():
    """Extract Hacker News stories using CSS selectors."""
    async with AsyncWebCrawler(api_key=API_KEY) as crawler:
        # Define CSS extraction schema
        config = CrawlerRunConfig(
            extraction_strategy={
                "type": "json_css",
                "schema": {
                    "name": "HackerNewsStories",
                    "baseSelector": ".athing",
                    "fields": [
                        {"name": "title", "selector": ".titleline > a", "type": "text"},
                        {"name": "url", "selector": ".titleline > a", "type": "attribute", "attribute": "href"},
                        {"name": "points", "selector": "+ tr .score", "type": "text"},
                        {"name": "author", "selector": "+ tr .hnuser", "type": "text"}
                    ]
                }
            }
        )

        print("Crawling Hacker News with CSS extraction...")
        result = await crawler.run(
            url="https://news.ycombinator.com",
            strategy="http",  # Fast, no browser needed
            config=config
        )

        if result.success and result.extracted_content:
            stories = json.loads(result.extracted_content)
            print(f"\nExtracted {len(stories)} stories")
            print("\nFirst 3 stories:")
            for story in stories[:3]:
                print(f"\nTitle: {story.get('title')}")
                print(f"URL: {story.get('url')}")
                print(f"Points: {story.get('points')}")
                print(f"Author: {story.get('author')}")
        else:
            print(f"Error: {result.error_message}")


if __name__ == "__main__":
    asyncio.run(extract_with_css())

File: nodejs/examples/extraction/01_css_extraction_sdk.ts
#!/usr/bin/env npx ts-node
/**
 * CSS Extraction with SDK - No LLM Cost
 *
 * This example shows how to extract structured data using CSS selectors.
 * CSS extraction is fast, reliable, and has no LLM cost.
 *
 * Usage:
 *     npx ts-node 01_css_extraction_sdk.ts
 *
 * Requirements:
 *     npm install crawl4ai-cloud
 */

import { AsyncWebCrawler } from 'crawl4ai-cloud';

// Configuration
const API_KEY = 'YOUR_API_KEY'; // Replace with your API key

async function extractWithCss(): Promise<void> {
  const crawler = new AsyncWebCrawler({ apiKey: API_KEY });

  try {
    // Define CSS extraction schema
    const config = {
      extraction_strategy: {
        type: 'json_css',
        schema: {
          name: 'HackerNewsStories',
          baseSelector: '.athing',
          fields: [
            { name: 'title', selector: '.titleline > a', type: 'text' },
            { name: 'url', selector: '.titleline > a', type: 'attribute', attribute: 'href' },
            { name: 'points', selector: '+ tr .score', type: 'text' },
            { name: 'author', selector: '+ tr .hnuser', type: 'text' },
          ],
        },
      },
    };

    console.log('Crawling Hacker News with CSS extraction...');
    const result = await crawler.run('https://news.ycombinator.com', {
      strategy: 'http', // Fast, no browser needed
      config,
    });

    if (result.success && result.extractedContent) {
      const stories = JSON.parse(result.extractedContent);
      console.log(`\nExtracted ${stories.length} stories`);
      console.log('\nFirst 3 stories:');
      stories.slice(0, 3).forEach((story: any) => {
        console.log(`\nTitle: ${story.title}`);
        console.log(`URL: ${story.url}`);
        console.log(`Points: ${story.points}`);
        console.log(`Author: ${story.author}`);
      });
    } else {
      console.log(`Error: ${result.errorMessage}`);
    }
  } finally {
    await crawler.close();
  }
}

extractWithCss().catch(console.error);
